src/crypto.ts

```diff
@@ -0,0 +1,427 @@
+ * asynchronously, this function returns a Promise<CryptoProvider>. The provider may be set using the `setProvider`
+ * function, which accepts a CryptoProvider.
+ */
+
+import { Codec, ByteBuffer } from 'dojo-core/encoding';
+import Promise from 'dojo-core/Promise';
+import { Sink } from 'dojo-core/streams/WritableStream';
+import has from './has';
+
+declare const define: { amd: any };
+declare const require: Function;
+
+export type Data = string | ByteBuffer;
+
+/**
+ * Supported hash algorithms 
+ */
+const HASH_ALGORITHMS = {
+	md5: true,
+	sha1: true,
+	sha256: true
+}
+
+/**
+ * Supported signing algorithms 
+ */
+const SIGN_ALGORITHMS = {
+	hmac: true
+}
+
+/**
+ * An interface describing a cryptographic provider.
+ */
+export interface CryptoProvider {
+	getHash(algorithm: string): HashFunction;
+	getSign(algorithm: string): SignFunction;
+}
+
+/**
+ * A function that can hash a chunk of data.
+ */
+export interface HashFunction {
+	(data: ByteBuffer): Promise<ByteBuffer>;
+	(data: string, codec?: Codec): Promise<ByteBuffer>;
+	create<T extends Data>(codec?: Codec): Hasher<T>;
+	algorithm: string;
+}
+
+/**
+ * A signing function.
+ */
+export interface SignFunction {
+	(key: Key, data: ByteBuffer): Promise<ByteBuffer>;
+	(key: Key, data: string, codec?: Codec): Promise<ByteBuffer>;
+	create<T extends Data>(key: Key, codec?: Codec): Signer<T>;
+	algorithm: string;
+}
+
+/**
+ * The current provider. Providers supply concrete implementations of the API described here. Users should not typically
+ * need to access providers directly.
+ */
+let provider: CryptoProvider;
+let providerPromise: Promise<CryptoProvider>;
+
+/**
+ * Gets the HashFunction for a particular algorithm. The algorithm is specified as a string for simplicity and
+ * extensibility.
+ */
+export function getHash(algorithm: string): HashFunction {
+	// If a provider has been loaded, defer to its getHash
+	if (provider) {
+		return provider.getHash(algorithm);
+	}
+
+	// Before a provider has been loaded, check whether the requested algorithm is one of the standard set. After a
+	// provider is loaded, it will handle algorithm verification itself.
+	if (!(algorithm in HASH_ALGORITHMS)) {
+		throw new Error('invalid algorithm; available algorithms are [ \'' +
+			Object.keys(HASH_ALGORITHMS).join('\', \'') + '\' ]');
+	}
+
+	let realHash: HashFunction;
+	const hashPromise = new Promise<HashFunction>(function (resolve, reject) {
+		getProvider().then(function (provider) {
+			realHash = provider.getHash(algorithm);
+			resolve(realHash);
+		}).catch(function (error) {
+			reject(error);
+		});
+	});
+
+	// Return a wrapper that will defer calls to the hash until a provider has been loaded.
+	const hashFunction = <HashFunction> function (data: Data, codec?: Codec): Promise<ByteBuffer> {
+		if (realHash) {
+			return realHash(<any> data, codec);
+		}
+		return hashPromise.then<ByteBuffer>(function (hash) {
+			return hash(<any> data, codec);
```

Comment:
does `data` need to be cast as `any` here? It looks like `hash` only accepts an argument of type `Data` anyway.